### PR TITLE
Fixes #3306 by introducing special purpose methods in TaskTracker

### LINKED ISSUE #3306: Resident Tasks: adjust task counts and take reservations into account

### PR DESCRIPTION
**Pull request by aquamatthias**
and use those instead of filtering results.

### CODE REVIEW

File: src/main/scala/mesosphere/marathon/core/launchqueue/LaunchQueue.scala [hunk 1]
```diff
@@ -24,6 +24,7 @@ object LaunchQueue {
       tasksLaunchedOrRunning: Int,
       backOffUntil: Timestamp) {
     def waiting: Boolean = tasksLeftToLaunch != 0 || taskLaunchesInFlight != 0
+    def launchedOrRunning: Int = tasksLaunchedOrRunning + taskLaunchesInFlight
```
Comment (meichstedt) [inline src/main/scala/mesosphere/marathon/core/launchqueue/LaunchQueue.scala]: Now there is a val `tasksLaunchedOrRunning` and a def `launchedOrRunning` – either use unambiguous names or get rid of one of them.

File: src/main/scala/mesosphere/marathon/MarathonSchedulerActor.scala [hunk 2]
```diff
@@ -514,14 +514,16 @@ class SchedulerActions(
...
-    val currentCount = taskTracker.appTasksSync(app.id).count(_.launched.isDefined)
+    val tasks = taskTracker.appTasksSync(app.id)
+    val runningCount = tasks.count(_.launched.isDefined)
+    val reserved = tasks.count(t => t.reservationWithVolumes.isDefined && t.launched.isEmpty)
+    val runningReserved = runningCount + reserved
     val targetCount = app.instances
 
-    if (targetCount > currentCount) {
-      log.info(s"Need to scale ${app.id} from $currentCount up to $targetCount instances")
+    if (targetCount > runningReserved) {
```
Comment (meichstedt) [inline src/main/scala/mesosphere/marathon/MarathonSchedulerActor.scala]: Not sure whether that's correct. Consider the following case:
```
running = 3
reserved = 0
app.instances = 3
```
No we scale down to `instances: 2`. Then we have
```
running = 2
reserved = 1
app.instances = 2
```
So that one task was previously running and is now stale. If we scale up again to ìnstances: 3`, the vals above would end up being
```
runningCount = 2
reserved = 1
runningReserved = 3
targetCount = 3
```
So `targetCount == runningReserved` which brings us into the last branch with `Not scaling`

File: src/main/scala/mesosphere/marathon/upgrade/DeploymentActor.scala [hunk 3]
```diff
@@ -116,7 +116,7 @@ private class DeploymentActor(
   }
 
   def scaleApp(app: AppDefinition, scaleTo: Int, toKill: Option[Iterable[Task]]): Future[Unit] = {
-    val runningTasks = taskTracker.appTasksSync(app.id).filter(_.launched.isDefined)
+    val runningTasks = taskTracker.appTasksLaunchedSync(app.id).filter(_.launched.isDefined)
```
Comment (meichstedt) [inline src/main/scala/mesosphere/marathon/upgrade/DeploymentActor.scala]: `appTasksLaunchedSync` already filters – you can remove `.filter(_.launched.isDefined)`

File: src/main/scala/mesosphere/marathon/core/task/tracker/TaskTracker.scala [hunk 4]
```diff
@@ -32,6 +33,7 @@ trait TaskTracker {
   def tasksByAppSync: TaskTracker.TasksByApp
   def tasksByApp()(implicit ec: ExecutionContext): Future[TaskTracker.TasksByApp]
 
+  def countRunningAppTasksSync(appId: PathId): Int
```
Comment (meichstedt) [inline src/main/scala/mesosphere/marathon/core/task/tracker/TaskTracker.scala]: rename to `...launched...`

File: src/main/scala/mesosphere/marathon/MarathonSchedulerActor.scala (hunk 2 above)
Comment (meichstedt) [inline src/main/scala/mesosphere/marathon/MarathonSchedulerActor.scala]: As discussed: maybe best if you revert these changes and we keep evaluating the taskTracker info based on `_.launched.isDefined ~ running` and check the launchQueue afterwards.

**Comment by meichstedt** [comment]
Minor comments plus the discussed issue in `MarathonSchedulerActor` – please rebase as well :)

File: src/main/scala/mesosphere/marathon/core/launchqueue/LaunchQueue.scala (hunk 1 above)
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/core/launchqueue/LaunchQueue.scala]: Very good point. 
Renamed: 
```
tasksLaunchedOrRunning -> tasksLaunched
launchedOrRunning  -> tasksLaunchedOrInFlight
```

so you have:
```
tasksLeftToLaunch
taskLaunchesInFlight
tasksLaunched
tasksLaunchedOrInFlight (shorthand for tasksLaunched + taskLaunchesInFlight)
```

File: src/main/scala/mesosphere/marathon/core/task/tracker/TaskTracker.scala (hunk 4 above)
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/core/task/tracker/TaskTracker.scala]: Done.

File: src/main/scala/mesosphere/marathon/upgrade/DeploymentActor.scala (hunk 3 above)
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/upgrade/DeploymentActor.scala]: Left over. Thanks.

**Comment by aquamatthias** [comment]
@meichstedt thanks for the review.
Addressed your comments in a separate commit.
Once you confirm on that, I can squash and rebase.

**Comment by meichstedt** [comment]
@aquamatthias LGTM – if you don't mind renaming `QueuedTaskCount` go ahead; anyways, please squash and rebase!

**Comment by meichstedt** [comment]
```
         |_               Awesome – thanks!
   _____|~ |____
  (  --         ~~~~--_,
   ~~~~~~~~~~~~~~~~~~~'`
```
